zerver/lib/actions.py

```diff
@@ -5912,3 +5912,17 @@ def do_delete_realm_export(user_profile: UserProfile, export: RealmAuditLog) ->
     export.extra_data = ujson.dumps(export_data)
     export.save(update_fields=['extra_data'])
     notify_realm_export(user_profile)
+
+def get_markdown_messages(user_profile: UserProfile, messages: List[int]) -> str:
+
+    buffer = []
+    for message_id in messages:
+        try:
+            (message, user_message) = access_message(user_profile, message_id)
```

Comment:
This should use `bulk_access_messages`.  While that's currently no better than what you have, it'll mean improving that function fixes this code path as well.

Reply:
Tim, this method has a param a sequence of Message:

bulk_access_messages(user_profile: UserProfile, messages: Sequence[Message]) -> List[Message]:

We have an int List.

It would be an additional process (create messages with the id or find messages by id).

Reply:
We could pass a generator to that existing function:

```python3
bulk_access_messages(user_profile, (access_message(user_profile, message_id)[0] for message_id in messages))
```

Or we could update the existing function to also take Message ID sequences:

```python3
def bulk_access_messages(user_profile: UserProfile, messages: Union[Sequence[Message], Sequence[int]]) -> List[Message]:
    filtered_messages = []

    for message_or_id in messages:
        if isinstance(message_or_id, Message):
            message = message_or_id
            user_message = get_usermessage_by_message_id(user_profile, message.id)
        elif isinstance(message_or_id, int):
            message, user_message = access_message(user_profile, message_id)
        else:
            raise TypeError('messages argument must be a Sequence[Message] or Sequence[int] of Message IDs')
            
        if has_message_access(user_profile, message, user_message):
            filtered_messages.append(message)
    return filtered_messages
```

I'm kinda meh about both approaches over the existing one, idk if there's a better way.

Reply:
`bulk_access_messages` has I think only one real caller and only needs the IDs, not the objects; it's already fetching the messages as `user_message.message` anyway in `get_usermessage_by_message_id`.  Just migrate its API as a prep commit.

(The main thing we use for this kind of application is `bulk_access_messages_expect_usermessage`, but that's not appropriate for this use case)